CRADLE/Normalize/normalize.py

```diff
@@ -167,107 +141,92 @@ def excludeOverlapRegion():
+		trainSetMeta.append((chromo, start, end, trainNumToSelect))
+
+	numProcess = min(len(trainSetMeta), commonVari.NUMPROCESS)
+	task = np.array_split(trainSetMeta, numProcess)
 	pool = multiprocessing.Pool(numProcess)
-	result = pool.map_async(getTrainSet, task).get()
+	trainSetResults = pool.map_async(getTrainSet, task).get()
 	pool.close()
 	pool.join()
 
 	trainSet = []
-	for i in range(len(result)):
-		trainSet.extend(result[i])
+	for result in trainSetResults:
+		trainSet.extend(result)
 
 	return trainSet
 
+
 def getTrainSet(subregions):
 	subTrainSet = []
 
-	for regionIdx in range(len(subregions)):
-		chromo = subregions[regionIdx][0]
-		start = int(subregions[regionIdx][1])
-		end = int(subregions[regionIdx][2])
-		trainNumToSelect = int(subregions[regionIdx][3])
-
-		if(trainNumToSelect == 0):
-			continue
+	for region in filter(lambda r: r[3] != 0, subregions):
+		chromo, start, end, trainNumToSelect = region
 
-		selectedStarts = random.sample(list(np.arange(start, end, TRAINBIN_SIZE)), trainNumToSelect)
-		for i in selectedStarts:
-			subTrainSet.append([chromo, i])
+		selectedStarts = sample(
+			list(np.arange(start, end, TRAINBIN_SIZE)), trainNumToSelect
+		)
+		for startIdx in selectedStarts:
+			subTrainSet.append((chromo, startIdx))
 
 	return subTrainSet
 
-def getScaler(trainSet):
-	global ob1Values
-	ob1 = pyBigWig.open(commonVari.CTRLBW_NAMES[0])
-
-	ob1Values = []
-	for i in range(len(trainSet)):
-		regionChromo = trainSet[i][0]
-		regionStart = int(trainSet[i][1])
-		regionEnd = regionStart + TRAINBIN_SIZE
-
-		temp = np.array(ob1.values(regionChromo, regionStart, regionEnd))
-		temp[np.isnan(temp) == True] = 0
-		ob1Values.extend(list(temp))
-	ob1.close()
-
-	task = []
-	sampleNum = len(commonVari.CTRLBW_NAMES) + len(commonVari.EXPBW_NAMES)
-	for i in range(1, sampleNum):
-		task.append([i, trainSet])
 
+def getScaler(trainSets):
+	# Passing ob1Values in to getScalerForEachSample via starmap_async is
+	# surprisingly slow and ~quadruples the running time of the whole utility,
+	# so its stays a global
+	global ob1Values
+	with pyBigWig.open(commonVari.CTRLBW_NAMES[0]) as ob1:
+		ob1Values = []
+		for trainSet in trainSets:
+			regionChromo, regionStart, _regionEnd, _overlap = trainSet
+			regionEnd = regionStart + TRAINBIN_SIZE
+
+			temp = np.array(ob1.values(regionChromo, regionStart, regionEnd))
+			temp[np.isnan(temp) == True] = 0
+			ob1Values.extend(list(temp))
+
+	sampleNum = commonVari.CTRLBW_NUM + commonVari.EXPBW_NUM
+	task = [(i, trainSets) for i in range(1, sampleNum)]
 	###### OBTAIN A SCALER FOR EACH SAMPLE
-	numProcess = len(task)
-	if(numProcess > commonVari.NUMPROCESS):
-		numProcess = commonVari.NUMPROCESS
+	numProcess = min(len(task), commonVari.NUMPROCESS)
 	pool = multiprocessing.Pool(numProcess)
-	scalerResult = pool.map_async(getScalerForEachSample, task).get()
+	scalerResult = pool.starmap_async(getScalerForEachSample, task).get()
```

Comment:
Sorry if I asked you this before. Could you let me know the difference of `map_async` and `startmap_async?`

Reply:
Sure! lets say I have the following code:
```python
def add_starmap(x, y):
    return x + y

def add_map(numbers):
    x, y = numbers
    return x + y

pool = multiprocessing.Pool(4)
tasks = [(1, 2), (3, 4), (5, 6), (7, 8)]
```

With `pool.map_async` I have to use `add_map`:

    pool.map_async(add_map, tasks)

and with `pool.starmap_async` I have to use `add_starmap`:

    pool.starmap_async(add_starmap, tasks)

So the difference is that `starmap` "unpacks" the values in a task into the arguments of the function, and `map` does not.

The reason it's called starmap is because "*" (a star) in python is an operator that "spreads" out the contents of the iterable it's being used on. For example:

```python
numbers = (1, 2)

# spread the numbers out across the arguments to add_starmap
add_starmap(*numbers)

# pass numbers in as one object
add_map(numbers)
```

Reply:
Got you. Thank you so much for the detailed explanation!